cmd/tools/grafana/grafana.go

```diff
@@ -34,57 +34,170 @@ const (
-	mode7Folder    Folder // Server-side Grafana folder name for 7-mode dashboards
-	datasource     string
-	variable       bool
-	client         *http.Client
-	headers        http.Header
-	config         string
-	prefix         string
-	useHttps       bool
-	useInsecureTLS bool
-	labels         []string
+	command             string // one of: import, export, clean
+	addr                string // URL of Grafana server (e.g. "http://localhost:3000")
+	token               string // API token issued by Grafana server
+	dir                 string
+	serverfolder        Folder
+	datasource          string
+	variable            bool
+	client              *http.Client
+	headers             http.Header
+	config              string
+	prefix              string
+	useHttps            bool
+	useInsecureTLS      bool
+	labels              []string
+	dirGrafanaFolderMap map[string]*Folder
 }
 
 type Folder struct {
-	folderName string // Grafana folder where to upload from where to download dashboards
-	folderId   int64
-	folderUid  string
+	name string // Grafana folder where to upload from where to download dashboards
+	id   int64
+	uid  string
+}
+
+func adjustOptions() {
+	opts.config = conf.ConfigPath(opts.config)
+	homePath = conf.GetHarvestHomePath()
+
+	// When opt.addr starts with https don't change it
+	if !strings.HasPrefix(opts.addr, "https://") {
+		opts.addr = strings.TrimPrefix(opts.addr, "http://")
+		opts.addr = strings.TrimPrefix(opts.addr, "https://")
+		opts.addr = strings.TrimSuffix(opts.addr, "/")
+		if opts.useHttps {
+			opts.addr = "https://" + opts.addr
+		} else {
+			opts.addr = "http://" + opts.addr
+		}
+	}
+}
+
+func askForToken() {
+	// ask for API token if not provided as arg and validate
+	if err := checkToken(opts, false); err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
 }
 
 func doExport(_ *cobra.Command, _ []string) {
 	adjustOptions()
+	validateExport()
 	askForToken()
-	var doesFolderExist = doesGrafanaFolderExist()
-	var err error
+	initExportVars()
+	fmt.Printf("preparing to export dashboards...\n")
+	exportDashboards(opts)
+}
+
+func validateExport() {
+	if opts.serverfolder.name == "" {
+		fmt.Printf("error: serverfolder is empty.\n")
+		os.Exit(1)
+	}
 
-	if !(doesFolderExist[&opts.cmodeFolder] && doesFolderExist[&opts.mode7Folder]) {
-		if !doesFolderExist[&opts.cmodeFolder] {
-			fmt.Printf("folder [%s] not found in Grafana\n", opts.cmodeFolder.folderName)
+	if opts.dir == "" {
+		fmt.Printf("error: dir is empty.\n")
+		os.Exit(1)
+	}
+
+	if opts.dir != "grafana/dashboards" {
+		exitIfExist(opts.dir, "directory")
+	}
+}
+
+func initExportVars() {
+	opts.dirGrafanaFolderMap = make(map[string]*Folder)
+
+	m := make(map[string]*Folder)
+
+	if opts.serverfolder.name != "" {
+		if opts.dir == "grafana/dashboards" {
+			m[path.Join(opts.dir, opts.serverfolder.name)] = &Folder{name: opts.serverfolder.name}
+		} else {
+			m[opts.dir] = &Folder{name: opts.serverfolder.name}
```

Comment:
don't we add serverfoldername ?

Reply:
do you mean why to have additional empty check of serverfoldername when it is mandatory?

Reply:
I thought some thing else, no comments here.